### ISSUE TITLE
non-deterministic bug, apparently in cleanStructuralRules

### ISSUE THREAD

**Issue by newca12** [newca12/gapt#264]
_From [stefan.hetzl](https://code.google.com/u/stefan.hetzl/) on April 04, 2014 17:42:17_

I have been experiencing the following strange behaviour. During 'mvn install', a test failed as below. I tried running this test again several times using 'mvn -pl integration_tests/misc_test install' and sometimes this problem did occur, and sometimes it did not (i.e. all tests passed successfully). In fact, after around 15 tries, the test seems to fail with a probability of only 20-25&#37;.

This looks to me like a bug in cleanStructuralRules which is sometimes exposed (and sometimes not) depending on the proof produced by reading back an expansion tree.

The bug should be fixed, even if it appears only sometimes.

Independently of that it would be nice to understand the cause of this non-deterministic behaviour. Bernhard, any thoughts on that?

\-------------------------------------------------------
 T E S T S
\-------------------------------------------------------
Running at.logic.integration_tests.MiscTest
Using OneVariableDelta.

End sequent: P(0), ∀x.(P(x)⊃P(s(x))) :- P(s(s(s(s(0)))))
Size of term set: 4
initializing generalized delta-table (set-based)

Number of grammars: 2
Smallest grammar-size: 4
Number of smallest grammars: 2
log4j:WARN No appenders could be found for logger (at.logic.algorithms.cutIntroduction.CutIntroduction$).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.

Grammar chosen: {List(tuple1(s(α_0)), tuple1(α_0))} o {Set(List(0), List(s(s(0))))}

Minimized cut formula: ∀x.(P(s(s(x)))∨¬P(x))

Warning: Axiom not uniquely specified, possible candidates:  A21   ∀x:ι.=(bm(x:ι, 0:ι):ι, x:ι): o, A20   ∀x:ι.=(bm(0:ι, x:ι):ι, 0:ι): o
Warning: Axiom not uniquely specified, possible candidates:  A35   ∀x:ι.∀y:ι.(<(+(x:ι, 1:ι):ι, +(y:ι, 1:ι):ι): o⊃<(x:ι, y:ι): o), A17   ∀x:ι.∀z:ι.∀y:ι.(<(+(x:ι, z:ι):ι, +(y:ι, z:ι):ι): o⊃<(x:ι, y:ι): o)
User specified sub works!{ Y  ->  (\lambda x ( - ((x + 1) \< x))) }
Warning: Axiom not uniquely specified, possible candidates:  IND   ∀Y:(ι->ο).((Y(0:ι): o∧∀n:ι.(Y(n:ι): o⊃Y(+(n:ι, 1:ι):ι): o))⊃∀n:ι.Y(n:ι): o), IND   ∀Y:(ι->ο).((Y(0:ι): o∧∀n:ι.(Y(n:ι): o⊃Y(+(n:ι, 1:ι):ι): o))⊃∀n:ι.Y(n:ι): o)
Warning: Axiom not uniquely specified, possible candidates:  A32   ∀x:ι.∀y:ι.(<(x:ι, y:ι): o⊃<(+(x:ι, 1:ι):ι, +(y:ι, 1:ι):ι): o), A36   ∀x:ι.∀z:ι.∀y:ι.(<(x:ι, y:ι): o⊃<(+(x:ι, z:ι):ι, +(y:ι, z:ι):ι): o)
User specified sub works!{ Y  ->  (\lambda x (x \< (x + 1))) }
Warning: Axiom not uniquely specified, possible candidates:  IND   ∀Y:(ι->ο).((Y(0:ι): o∧∀n:ι.(Y(n:ι): o⊃Y(+(n:ι, 1:ι):ι): o))⊃∀n:ι.Y(n:ι): o), IND   ∀Y:(ι->ο).((Y(0:ι): o∧∀n:ι.(Y(n:ι): o⊃Y(+(n:ι, 1:ι):ι): o))⊃∀n:ι.Y(n:ι): o)
wrong sub found! ((exists x_{16} (MON(h, 0) & NOCC(h, 0, \sigma))) & (all n ((exists x_{17} (MON(h, n) & NOCC(h, n, \sigma))) -> (exists x_{18} (MON(h, (n + 1)) & NOCC(h, (n + 1), \sigma)))))) -> (all n (exists x_{19} (MON(h, n) & NOCC(h, n, \sigma))))  for  ((exists h (MON(h, 0) & NOCC(h, 0, \sigma))) & (all n ((exists h (MON(h, n) & NOCC(h, n, \sigma))) -> (exists h (MON(h, (n + 1)) & NOCC(h, (n + 1), \sigma)))))) -> (all n (exists h (MON(h, n) & NOCC(h, n, \sigma))))  sub={ h  ->  h } ax= ((exists h (MON(h, 0) & NOCC(h, 0, \sigma))) & (all n ((exists h (MON(h, n) & NOCC(h, n, \sigma))) -> (exists h (MON(h, (n + 1)) & NOCC(h, (n + 1), \sigma)))))) -> (all n (exists h (MON(h, n) & NOCC(h, n, \sigma))))
User specified sub works!{ Y  ->  (\lambda x (exists h (MON(h, x) & NOCC(h, x, \sigma)))) }
Tests run: 12, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 8.468 sec \<<< FAILURE! - in at.logic.integration_tests.MiscTest
The system should::Construct proof with expansion sequent extracted from proof 2/2(at.logic.integration_tests.MiscTest)  Time elapsed: 0.11 sec  \<<< ERROR!
at.logic.calculi.lk.base.LKUnaryRuleCreationException: Could not create lk rule c:l from parent P(0), (P(0)⊃P(s(0))) :- P(s(0)) with auxiliary formulas P(0)
    at at.logic.calculi.lk.propositionalRules.ContractionLeftRule$.apply(propositionalRules.scala:266)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:49)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:53)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:53)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:234)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:53)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:114)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:114)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:114)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:114)
    at at.logic.algorithms.lk.CleanStructuralRules$.cleanStructuralRules(elimination.scala:43)
    at at.logic.algorithms.lk.CleanStructuralRules$.apply(elimination.scala:23)
    at at.logic.algorithms.lk.solve$.startProving(solve.scala:66)
    at at.logic.algorithms.lk.solve$.expansionProofToLKProof(solve.scala:55)
    at at.logic.integration_tests.MiscTest$$anonfun$1$$anonfun$apply$42.apply(MiscTest.scala:259)
    at at.logic.integration_tests.MiscTest$$anonfun$1$$anonfun$apply$42.apply(MiscTest.scala:255)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire...

_Original issue: http://code.google.com/p/gapt/issues/detail?id=264_

**Comment by newca12** [comment]
_From [giselle....@gmail.com](https://code.google.com/u/108711232698608960044/) on April 04, 2014 10:01:06_

I only tested 4 times before committing... damn!! :D

So I can reproduce the error here, and I observed the following:
prove(seq_norm, strategy) returns different proofs every call. I checked
the number of contractions, since this was the rule causing problems, and
for pairs (#c_l, #c_r), the tests run smoothly for: (24, 4); (27, 7); (20,
8); (24, 10); (21, 7)
And they fail for: (35, 11); (32, 14)

Is this the intended behavior? I mean, that the procedure returns different
proofs with the same strategy?
